### PR TITLE
Add `is_valid_letter()` to `TextServer`

### LINKED ISSUE #7656: Add `TextServer.is_valid_letter(unicode:int)`
### Describe the project you are working on

A game where the player can type text, which is interpreted and checked against a known vocabulary and grammar.

### Describe the problem or limitation you are having in your project

When the user writes text into TextEdit, we pre-check if the input is valid
* letters
* punctuation: ? , . ! -> easy enough as there is a tiny amount of valid options
* potentially digits ([quasi-int](https://docs.godotengine.org/en/latest/classes/class_string.html#class-string-method-is-valid-int), where I might need to strip `+`/`-`) -> though an `TextServer.is_valid_digit` with no prefix would be nicer

Now then, checking for `is letter` becomes increasingly difficult if you leave the ASCII space:  
You can get an [unicode](https://docs.godotengine.org/en/stable/classes/class_inputeventkey.html#class-inputeventkey-property-unicode), but with all the diverse languages, it is difficult to judge if the input is actually a letter

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

What we need is an equivalent to [is_valid_identifier](https://docs.godotengine.org/en/latest/classes/class_textserver.html#class-textserver-method-is-valid-identifier).

Fortunately, Godot already uses icu4c. And it can check if the unicode code point is part of the general categories "L" (letters):

https://github.com/godotengine/godot/blob/master/thirdparty/icu4c/common/unicode/uchar.h#L3062

Better an `int` than `String` -> [font_has_char](https://docs.godotengine.org/en/latest/classes/class_textserver.html#class-textserver-method-font-has-char). And we can always easily change it to String with `String.chr()`.
```c++
TextServer::is_valid_letter(int unicode);
```

As we simply want to know whether we have a (natural language) letter or not, we probably don't need the locale information: https://heistak.github.io/your-code-displays-japanese-wrong/
Unicode code points are not really… unique. `isalpha` seems to have an abstraction to fix that, requiring locale info: https://cplusplus.com/reference/locale/isalpha/

> **Warning**
There's only one thing one has to pay attention to: _`String.chr()` may render the wrong letter_. Choosing a font that only supports japanese or chinese or etc. is a workaround, see https://github.com/godotengine/godot-proposals/issues/3766#issuecomment-1006612914

> **Note**
An alternative to `u_isalpha()` would be to use `u_hasBinaryProperty(UChar32 c, UProperty which)` for advanced checks (but we would be fine with the letter check):

https://github.com/godotengine/godot/blob/3ed4497113fa10611b90290ce22a751fb9d26e2e/thirdparty/icu4c/common/unicode/uchar.h#L2696

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

```gdscript
var valid_input:String

func _gui_input(event: InputEvent) -> void:
  if event is InputEventKey && event.is_pressed():
   var uni:int = event.get_unicode()
    var ch:String = String.chr(uni)
    if ch.length() > 0 && (TextServer.is_valid_letter(uni)  || _is_valid_punctuation(ch)):
      valid_input += ch

func _is_valid_punctuation(ch:String) -> bool:
  pass
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Not a few lines. One needs to include icu4c. And GDScript users lose out.

### Is there a reason why this should be core and not an add-on in the asset library?

Godot already includes icu4c.

### PR DESCRIPTION
**Pull request by MJacred**
Fixes https://github.com/godotengine/godot-proposals/issues/7656

Test project:
[letter_test.tar.gz](https://github.com/godotengine/godot/files/12607718/letter_test.tar.gz)

related PR:
* https://github.com/godotengine/godot/pull/75846

### CODE REVIEW

**Comment by MJacred** [comment]
The code works, but the list of ranges is not complete. I used https://www.unicode.org/Public/UCD/latest/ucd/PropList.txt, but it's  only a subset. https://unicode.org/Public/UNIDATA/DerivedCoreProperties.txt seems to be the correct file for this…

**Comment by MJacred** [comment]
@adamscott: ping! you wanted to review this PR?

**Comment by MJacred** [comment]
@adamscott: reminder of this PR

File: tests/servers/test_text_server.h [hunk 1]
```diff
@@ -637,6 +637,36 @@ TEST_SUITE("[TextServer]") {
...
 		}
 
+		SUBCASE("[TextServer] Unicode letters") {
+			for (int i = 0; i < TextServerManager::get_singleton()->get_interface_count(); i++) {
+				Ref<TextServer> ts = TextServerManager::get_singleton()->get_interface(i);
+				CHECK_FALSE_MESSAGE(ts.is_null(), "Invalid TS interface.");
+
+				static const char32_t *words[19] = { U"-30", U"100", U"10.1", U"10,1", U"1e2", U"1e-2", U"1e2e3", U"0xAB", U"AB", U"Test1", U"1Test", U"Test*1", U"test_testeT", U"test_tes teT", U"\u0643\u0402\u0716\u0b05", U"\u0643\u0402\u0716\u0b05Test\u30aa\u4e21", U"ӒӖӚӜ", U"_test", U"ÂÃÄÅĀĂĄÇĆĈĊ" };
+				static bool words_with_just_letters[19] = { false, false, false, false, false, false, false, false, true, false, false, false, false, false, true, true, true, false, true };
+				for (int j = 0; j < 19; j++) {
```
Comment (dalexeev) [inline tests/servers/test_text_server.h:647]: Perhaps it would be more convenient to use a struct and write test cases in a column?

File: tests/servers/test_text_server.h [hunk 2]
```diff
@@ -637,6 +637,36 @@ TEST_SUITE("[TextServer]") {
...
+					String word = String(words[j]);
+					bool contains_just_letters = true;
+					bool allow_only_letters = words_with_just_letters[j];
+					int failed_on_index = -1;
+					for (int k = 0; k < word.length(); k++) {
+						if (!ts->is_valid_letter(word[k])) {
+							contains_just_letters = false;
+							if (allow_only_letters) { // early exit.
```
Comment (dalexeev) [inline tests/servers/test_text_server.h:655]:
```suggestion
							if (allow_only_letters) { // Early exit.
```

File: modules/text_server_adv/text_server_adv.cpp [hunk 3]
```diff
@@ -7229,6 +7229,17 @@ bool TextServerAdvanced::_is_valid_identifier(const String &p_string) const {
 	return true;
 }
 
+bool TextServerAdvanced::_is_valid_letter(char32_t p_unicode) const {
+#ifndef ICU_STATIC_DATA
+	if (!icu_data_loaded) {
+		WARN_PRINT_ONCE("ICU data is not loaded, Unicode security and spoofing detection disabled.");
```
Comment (bruvzg) [inline modules/text_server_adv/text_server_adv.cpp:7235]: This is not related to spoofing and security, not sure if any warning is necessary.

**Review by dalexeev — COMMENTED** [review]
Overall looks good to me, but I'm not a Unicode expert. @bruvzg Could you please double check?

**Review by bruvzg — APPROVED** [review]

**Review by AThousandShips — APPROVED** [review]

File: modules/text_server_adv/text_server_adv.cpp (hunk 3 above)
Comment (MJacred) [reply modules/text_server_adv/text_server_adv.cpp:7235]: Sure, I can remove it. Copied it from `TextServerAdvanced::_is_valid_identifier` just in case

File: tests/servers/test_text_server.h (hunk 1 above)
Comment (MJacred) [reply tests/servers/test_text_server.h]: Ok, I rewrote the test, please have a look.  
Sadly, I couldn't run the tests locally because of some errors that look like they don't come from my code

<details>

<summary>Backtrace</summary>

```
./godot.linuxbsd.editor.x86_64.llvm --test
[doctest] doctest version is "2.4.11"
[doctest] run with "--help" for options
Could not load project settings.
ERROR: Could not open specified test directory.
   at: make_tests (modules/gdscript/tests/gdscript_test_runner.cpp:326)
===============================================================================
./modules/gdscript/tests/gdscript_test_runner_suite.h:43:
TEST SUITE: [Modules][GDScript]
TEST CASE:  Script compilation and runtime

modules/gdscript/tests/gdscript_test_runner.cpp:191: FATAL ERROR: An error occurred while making the tests.

./modules/gdscript/tests/gdscript_test_runner_suite.h:49: FATAL ERROR: REQUIRE( fail_count == 0 ) is NOT correct!
  values: REQUIRE( -1 == 0 )
  logged: Make sure `*.out` files have expected results.
          All GDScript tests should pass.

ERROR: Could not create directory: res://.godot
   at: make_dir_recursive (core/io/dir_access.cpp:180)
ERROR: Could not create project data directory (res://.godot) in: res://
   at: EditorPaths (editor/editor_paths.cpp:235)
ERROR: Failed to create file res://.godot/.gdignore
   at: EditorPaths (editor/editor_paths.cpp:248)
Could not load project settings.
===============================================================================
./modules/gdscript/tests/test_completion.h:191:
TEST SUITE: [Modules][GDScript][Completion]
TEST CASE:  [Editor] Check suggestion list

./modules/gdscript/tests/test_completion.h:86: FATAL ERROR: Invalid test directory.

ERROR: Can't save resource to empty path. Provide non-empty path or a Resource with non-empty resource_path.
   at: save (core/io/resource_saver.cpp:105)
ERROR: Error saving editor settings to 
   at: save (editor/editor_settings.cpp:1108)
===============================================================================
./modules/gdscript/tests/test_lsp.h:391:
TEST SUITE: [Modules][GDScript][LSP]
TEST CASE:  [workspace][resolve_symbol]

./modules/gdscript/tests/test_lsp.h:90: FATAL ERROR: REQUIRE( err == OK ) is NOT correct!
  values: REQUIRE( 31 == 0 )
  logged: Could not open specified root directory

./modules/gdscript/tests/test_lsp.h:391: FATAL ERROR: test case CRASHED: SIGSEGV - Segmentation violation signal

===============================================================================
[doctest] test cases:  6 |  3 passed | 3 failed | 866 skipped
[doctest] assertions: 53 | 49 passed | 4 failed |
[doctest] Status: FAILURE!

================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.3.dev.custom_build (23191b834e4609baacf19855c3acb6a9f607b30f)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] /lib/x86_64-linux-gnu/libpthread.so.0(+0x14420) [0x7fb31acc7420] (??:0)
[2] ./godot.linuxbsd.editor.x86_64.llvm() [0x2a4144e] (??:0)
[3] ./godot.linuxbsd.editor.x86_64.llvm() [0x2a4380a] (??:0)
[4] ./godot.linuxbsd.editor.x86_64.llvm() [0x2b6e25a] (??:0)
[5] ./godot.linuxbsd.editor.x86_64.llvm() [0x2a4499e] (??:0)
[6] ./godot.linuxbsd.editor.x86_64.llvm() [0x26a1645] (??:0)
[7] ./godot.linuxbsd.editor.x86_64.llvm() [0x261b5b0] (??:0)
[8] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7fb31a990083] (??:0)
[9] ./godot.linuxbsd.editor.x86_64.llvm() [0x261b4ae] (??:0)
-- END OF BACKTRACE --
================================================================
Aborted (core dumped)

```

</details>

Comment (bruvzg) [reply tests/servers/test_text_server.h]: > Sadly, I couldn't run the tests locally because of some errors that look like they don't come from my code

Test should be run from the repo root not `bin`, some rely on current working directory to load stuff.

Comment (MJacred) [reply tests/servers/test_text_server.h]: Argh, I forgot about that. Thanks!  
So, I ran the test and it failed (will look into it)

Comment (MJacred) [reply tests/servers/test_text_server.h]: OK, found and fixed the issue in my test

**Comment by akien-mga** [comment]
Thanks!
